mailbox/store/src/main/java/org/apache/james/mailbox/store/StoreMessageIdManager.java

```diff
@@ -0,0 +1,252 @@
+        for (Map.Entry<MailboxId, UpdatedFlags> entry : updatedFlags.entrySet()) {
+            dispatchFlagsChange(mailboxSession, entry.getKey(), entry.getValue());
+        }
+    }
+
+    @Override
+    public List<MessageResult> getMessages(List<MessageId> messageIds, final MessageResult.FetchGroup fetchGroup, MailboxSession mailboxSession) throws MailboxException {
+        try {
+            MessageIdMapper messageIdMapper = mailboxSessionMapperFactory.getMessageIdMapper(mailboxSession);
+            return FluentIterable.from(messageIdMapper.find(messageIds, MessageMapper.FetchType.Full))
+                .transform(messageResultConverter(fetchGroup))
+                .toList();
+        } catch (WrappedException wrappedException) {
+            throw wrappedException.unwrap();
+        }
+    }
+
+    @Override
+    public void delete(MessageId messageId, final List<MailboxId> mailboxIds, MailboxSession mailboxSession) throws MailboxException {
+        MessageIdMapper messageIdMapper = mailboxSessionMapperFactory.getMessageIdMapper(mailboxSession);
+        MailboxMapper mailboxMapper = mailboxSessionMapperFactory.getMailboxMapper(mailboxSession);
+
+        Iterable<MetadataWithMailboxId> metadatasWithMailbox = FluentIterable
+            .from(messageIdMapper.find(ImmutableList.of(messageId), MessageMapper.FetchType.Metadata))
+            .filter(inMailbox(mailboxIds))
+            .transform(EXTRACT_METADATA_FUNCTION);
+
+        messageIdMapper.delete(messageId, mailboxIds);
+
+        for (MetadataWithMailboxId metadataWithMailboxId : metadatasWithMailbox) {
+            dispatcher.expunged(mailboxSession, metadataWithMailboxId.messageMetaData, mailboxMapper.findMailboxById(metadataWithMailboxId.mailboxId));
+        }
+    }
+
+    @Override
+    public void setInMailboxes(MessageId messageId, List<MailboxId> mailboxIds, MailboxSession mailboxSession) throws MailboxException {
+        MessageIdMapper messageIdMapper = mailboxSessionMapperFactory.getMessageIdMapper(mailboxSession);
+
+        List<MailboxMessage> mailboxMessages = messageIdMapper.find(ImmutableList.of(messageId), MessageMapper.FetchType.Full);
+
+        if (!mailboxMessages.isEmpty()) {
+            ImmutableSet<MailboxId> currentMailboxes = FluentIterable.from(mailboxMessages)
+                .transform(EXTRACT_MAILBOX_ID_FUNCTION)
+                .toSet();
+            HashSet<MailboxId> targetMailboxes = Sets.newHashSet(mailboxIds);
+            List<MailboxId> mailboxesToRemove = ImmutableList.copyOf(Sets.difference(currentMailboxes, targetMailboxes));
+
+            SetView<MailboxId> mailboxesToAdd = Sets.difference(targetMailboxes, currentMailboxes);
+
+            if (!mailboxesToAdd.isEmpty()) {
+                MailboxMessage mailboxMessage = mailboxMessages.get(0);
+
+                validateQuota(mailboxIds, mailboxSession, mailboxMessage);
```

Comment:
You should also decrement the count you check for each deleted message.

I would write outside of the if : 

```
validateQuota(mailboxIds, mailboxSession, mailboxMessage, mailboxesToAdd.size() - mailboxesToRemove.size());

if (!mailboxesToAdd.empty()) {
    ...
}
if (!mailboxesToRemove.empty()) {
    ...
}
```